src/library_manager/lib_manager.c

```diff
@@ -0,0 +1,455 @@
+			DMA_ACCESS_EXCLUSIVE);
+	if (!dma_ext->dma.dmac) {
+		tr_err(&lib_manager_tr, "dma_ext_init(): dma.dmac = NULL");
+		return -ENODEV;
+	}
+
+	/* get required address alignment for dma buffer */
+	ret = dma_get_attribute(dma_ext->dma.dmac, DMA_ATTR_BUFFER_ADDRESS_ALIGNMENT,
+					&addr_align);
+	if (ret < 0) {
+		tr_err(&lib_manager_tr, "dma_get_attribute(): ret = %d", ret);
+		return ret;
+	}
+
+	dma_ext->chan = dma_channel_get(dma_ext->dma.dmac, dma_id);
+	if (!dma_ext->chan) {
+		tr_err(&lib_manager_tr, "dma_channel_get(): dma_ext->chan = NULL");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**
+ * \brief Stop, deinit and free DMA and buffer used by loader.
+ *
+ * \return 0 on success, error code otherwise.
+ */
+
+static int lib_manager_dma_deinit(struct lib_manager_dma_ext *dma_ext)
+{
+	dma_channel_put(dma_ext->chan);
+	dma_put(dma_ext->dma.dmac);
+	return 0;
+}
+
+static int lib_manager_load_data_from_host(struct lib_manager_dma_ext *dma_ext, uint32_t size)
+{
+	struct dma_sg_config config;
+	uint32_t elem_addr;
+	uint32_t elem_num, elem_size;
+	int ret = -ENOMEM;
+	uint32_t free_bytes = 0;
+	uint32_t avail_bytes = 0;
+
+	elem_addr = (uint32_t)dma_ext->dmabp.addr;
+
+	elem_size = sizeof(uint64_t) * DMA_ELEM_SIZE;
+	elem_num = size / elem_size;
+
+	config.direction = DMA_DIR_HMEM_TO_LMEM;
+	config.src_width = sizeof(uint32_t);
+	config.dest_width = sizeof(uint32_t);
+	config.cyclic = 0;
+
+	ret = dma_sg_alloc(&config.elem_array, SOF_MEM_ZONE_RUNTIME,
+			   config.direction, elem_num, elem_size, elem_addr, 0);
+
+	if (ret < 0)
+		return ret;
+
+	ret = dma_set_config(dma_ext->chan, &config);
+	if (ret < 0)
+		return ret;
+
+	ret = dma_start(dma_ext->chan);
+	if (ret < 0)
+		return ret;
+
+	/* Wait till whole data acquired */
+	while (avail_bytes < size) {
+		/* get data sizes from DMA */
+		ret = dma_get_data_size(dma_ext->chan, &avail_bytes, &free_bytes);
+		if (ret < 0)
+			return ret;
+
+		/* Exit if avail_bytes == 0 */
+		if (avail_bytes == 0)
+			break;
+#ifndef __ZEPHYR__
+		wait_delay(1);
+#else
+		Z_TIMEOUT_CYC(1);
+#endif
+		if (ret < 0)
+			return ret;
+	}
+
+	ret = dma_stop_delayed(dma_ext->chan);
+	if (ret < 0)
+		return ret;
+
+	dma_sg_free(&config.elem_array);
+
+	dcache_invalidate_region((void *)dma_ext->dmabp.addr, size);
+
+	return ret;
+}
+
+/* TODO
+ * This is stub, that will be replaced when TLB support
+ * will be available.
+ */
+static int mm_map(uint32_t *address, uint32_t size, uint32_t attribs)
+{
+	uint32_t addr_align = HOST_PAGE_SIZE;
+	void *local_add;
+	/* allocate new buffer */
+	local_add = (void *)rballoc_align(0, SOF_MEM_CAPS_DMA,
+						size, addr_align);
+	tr_err(&lib_manager_tr, "mm_map() : pointer: 0x%x", local_add);
+	if (!address) {
+		tr_err(&lib_manager_tr, "dma_buffer_init(): alloc failed");
+		return -ENOMEM;
+	}
+
+	*address = (uint32_t)local_add;
+	return 0;
+}
+
+static int lib_manager_store_library(struct lib_manager_dma_ext *dma_ext, uint32_t lib_id,
```

Comment:
best to have a some function comments on some of the bigger static functions so everyone can follow.